### DISCUSSION TITLE
Sampling Accelerometer and Gyroscope Data When Screen is Off

### DISCUSSION THREAD

**Discussion by Delayless** [zepp-health/zeppos-forum#370] [Q&A]
I want to periodically collect accelerometer (e.g., once every 500ms) and gyroscope data for developing a application. I use a timer for sampling, but I found that sampling stops after the screen turns off (The Page Object call onDestroy first, and after a while, the App Object 
 also executes its‘ onDestroy. When waking up again, it returns to the Home screen).  I previously considered using AppService for this, but I found that the timer and gyroscope data are not available. Is there any way to handle this scenario?

**Comment by silver-zepp** [comment]
Currently, AOD should be the best solution for this task.

**Reply by Delayless** [reply]
> Currently, AOD should be the best solution for this task.

I tried the method you mentioned that set to AOD, but it still doesn't work. My device is the T-Rex Ultra with latest system version [zeppos3.5 ](version: 9.2.4.1)

The following are the logs with the duplicate sensor data omitted.
```

info [DEBUG-02:07:16.434] jsapp:20001 [I] [FreeStyle] [jsfwk.info ] [calories-app] [-] acc onInit 
info [DEBUG-02:07:16.473] jsapp:20001 [I] [FreeStyle] [jsfwk.info ] [calories] [-] build Page 
info [DEBUG-02:07:16.474] jsapp:20001 [I] [FreeStyle] [jsfwk.info ] [calories-app] [-] app.js onCreate foodType: chocolate 
info [DEBUG-02:07:16.729] jsapp:20001 [I] [FreeStyle] [jsfwk.info ] [calories-app] [-] 500ms: ACC: {"x":198.486328125,"y":63.232421875,"z":976.318359375} 
info [DEBUG-02:07:17.223] jsapp:20001 [I] [FreeStyle] [jsfwk.info ] [calories-app] [-] 500ms: ACC: {"x":182.373046875,"y":-69.3359375,"z":995.849609375} 
............ 
info [DEBUG-02:07:26.317] jsapp:20001 [I] [FreeStyle] [jsfwk.info ] [calories-app] [-] 500ms: ACC: {"x":122.55859375,"y":-879.638671875,"z":478.759765625}
info [DEBUG-02:07:26.819] jsapp:11611 [I] index page.js on init invoke 
info [DEBUG-02:07:26.820] jsapp:11611 [I] 79 
info [DEBUG-02:07:26.820] jsapp:11611 [I] ----->>>current
info [DEBUG-02:07:26.820] jsapp:11611 [I] [object Object]
info [DEBUG-02:07:26.854] jsapp:20001 [I] [FreeStyle] [jsfwk.info ] [calories-app] [-] 500ms: ACC: {"x":120.1171875,"y":-868.65234375,"z":505.126953125} 
............
info [DEBUG-02:07:31.365] jsapp:11611 [I] index page.js on destory invoke 
info [DEBUG-02:07:31.406] jsapp:20001 [I] [FreeStyle] [2024-09-24T02:07:32.876Z] [jsfwk.info ] [calories-app] [-] 500ms: ACC: {"x":173.33984375,"y":25.390625,"z":976.806640625} 
............
info [DEBUG-02:07:41.575] jsapp:11611 [I] ----->>>current 
info [DEBUG-02:07:41.576] jsapp:11611 [I] index page.js on init invoke 
info [DEBUG-02:07:41.576] jsapp:11611 [I] 80
info [DEBUG-02:07:41.576] jsapp:11611 [I] [object Object]
info [DEBUG-02:07:41.589] jsapp:20001 [I] [FreeStyle] [2024-09-24T02:07:43.092Z] [jsfwk.info ] [calories-app] [-] 500ms: ACC: {"x":185.05859375,"y":5.859375,"z":978.02734375} 
............
info [DEBUG-02:07:51.732] jsapp:20001 [I] [FreeStyle] [2024-09-24T02:07:53.232Z] [jsfwk.info ] [calories-app] [-] app.js onDestroy 
info [DEBUG-02:07:51.733] jsapp:20001 [I] [FreeStyle] [2024-09-24T02:07:53.223Z] [jsfwk.info ] [calories] [-] Page Destroyed 
```

**Reply by silver-zepp** [reply]
Sorry, I wasn't clear. You have to use AOD for the purpose of running your own watch face (instead of an app). This way you can constantly execute your code compared to a mini app that will have to be destroyed sooner or later.

**Reply by Delayless** [reply]
> Sorry, I wasn't clear. You have to use AOD for the purpose of running your own watch face (instead of an app). This way you can constantly execute your code compared to a mini app that will have to be destroyed sooner or later.

Thank you for your help; I tried the solution you suggested. The watchface can collect data at intervals, but I checked the available APIs and found that they do not provide the accelerometer and gyroscope data for watchface.
Is there a way to solve this scenario? Additionally, designed a watchface that is collecting sensor data is not very suitable for developing a device app, as the app will be coupled with the watchface.(of course, it doesn't matter for me, as long as the problem is solved.)

**Comment by silver-zepp** [comment]
Make sure your [watch face is using 3.0 API](https://github.com/zepp-health/zeppos-samples/tree/main/watchface/3.0/timer) and then refer to these APIs: 
1) http://docs.zepp.com/docs/reference/device-app-api/newAPI/sensor/Accelerometer
2) http://docs.zepp.com/docs/reference/device-app-api/newAPI/sensor/Gyroscope

**Reply by Delayless** [reply]
> Make sure your [watch face is using 3.0 API](https://github.com/zepp-health/zeppos-samples/tree/main/watchface/3.0/timer) and then refer to these APIs:
> 
> 1. http://docs.zepp.com/docs/reference/device-app-api/newAPI/sensor/Accelerometer
> 2. http://docs.zepp.com/docs/reference/device-app-api/newAPI/sensor/Gyroscope

Thank you. these sensor are worked in API3.0 watchface. But I found that some others APIs of the watchface are unavailable. For example, in the official 3.0 version watchface example `zeppos-samples\watchface\3.0\timer`, using this API [SysProSetInt](https://docs.zepp.com/zh-cn/docs/watchface/api/hmFS/SysProSetInt/) throws an undefined error.
```
I>0>51.968 quickjs- SBEngine [Timer] [2024-09-25T09:07:51.965Z] [jsfwk.error] [timer-page] [-] set t0 error: {}
I>0>51.969 quickjs- SBEngine TypeError: cannot read property 'SysProSetInt' of undefined
I>0>51.969 quickjs- SBEngine     at <anonymous> (/storage/data/js_watchfaces/00004E21/watchface/gtr-3-pro/index.js:322)
I>0>52.568 quickjs- SBEngine TypeError: cannot read property 'stopTimer' of undefined
I>0>52.568 quickjs- SBEngine     at <anonymous> (/storage/data/js_watchfaces/00004E21/watchface/gtr-3-pro/index.js:293)
```

**Reply by silver-zepp** [reply]
For 3.0, make sure you refer to `App` instead of `Watchface` API on our website, as it's no longer maintained. You have at least two options here:
1. [Easy Storage](https://github.com/silver-zepp/zeppos-easy-storage) library (plenty of storage solutions)
   - if you want a replacement for SysPro... you can use [EasyTempStorage](https://github.com/silver-zepp/zeppos-easy-storage/blob/master/docs/easy-temp-storage.md)
   - if you want permanent storage you use [EasyStorage](https://github.com/silver-zepp/zeppos-easy-storage/blob/master/docs/easy-storage.md) or other solutions like time-series database [EasyTSDB](https://github.com/silver-zepp/zeppos-easy-storage/blob/master/docs/easy-tsdb.md), which might be the best suit to your problem
2. [LocalStorage](https://docs.zepp.com/docs/reference/device-app-api/newAPI/storage/localStorage) (zos 2x+ built-in solution)

This is a huge difference between 1.0 and 3.0, the ability to import external npm libraries.

**Reply by silver-zepp** [reply]
And for your other timer-related issues, make sure to use these JS built-in methods: https://javascript.info/settimeout-setinterval, instead of createTimer/stopTimer; those are deprecated.

**Reply by Delayless** [reply]
Got it. Your team's new development logic for 3.0 is to provide necessary hardware interface calls and to maintain compatibility with native JavaScript and its extensibility, while also reducing the differences between watchface and app development, right?
I remember seeing in the docs that some high-power sensors' permission would be limited for watchfaces. Based on your current development logic(if what i said is correct), has this restriction been removed?

**Reply by silver-zepp** [reply]
The restrictions you refer to are for [background service](http://docs.zepp.com/docs/guides/framework/device/app-service/), in particular, hence the use of AOD.
